### PR TITLE
fix(core/Cardano): do not show change output in byron-shelley transfers

### PR DESCRIPTION
**Pull request by refi93**
Motivation: @tsusanka and Trezor QA noticed that Trezor shows the base address change output if some of the inputs happens to be a Byron era input. This is the result of the current `_should_hide_output()` logic in Cardano which hides outputs only if their path matches with all the inputs up to the account index which is stricter than it needs to be and causes issues when Byron/Shelley paths otherwise belonging to the same logical account are mixed.

Changes: relax the `_should_hide_output()` logic to look only at the account indices when matching the tx output with the inputs. The path roots (`m/1852'/1815'/` for Shelley and `m/44'/1815'/` for Byron) are being enforced by the Keychain (core/src/apps/cardano/seed.py) already. Also sending funds the other way around, i.e. from shelley to byron addresses is already independently accompanied by a warning that given change address has no stake rights (so users don't accidentally "unstake" their funds)

Testing:

* added testcase with a tx with a byron input and shelley output which fails with the previous logic as one more screen appeared, requiring a different input flow
* `pytest tests/device_tests -k "test_cardano_sign_tx"` passes

### CODE REVIEW

**Comment by tsusanka** [comment]
To avoid confusion: we have agreed with @refi93 this is an edge-case so we are not hurrying this into the upcoming release, we will release it next time.

Thanks @refi93 for the fix and :clap: to @bosomt for finding this.

**Comment by matejcik** [comment]
please rebase on current master and fix any typing errors

**Comment by refi93** [comment]
@matejcik done

**Comment by matejcik** [comment]
...and please fix conflicts again, this breaks on top of "allow more accounts" patch.

I'm not sure how to resolve this without endlessly creating more work for you :( Any ideas?

**Comment by matejcik** [comment]
i mean, i _could_ ask you to rebase the PRs in order of processing, as opposed to all at the same time :) next time then.

**Review by matejcik — APPROVED** [review]
otherwise LGTM and ready to merge

**Comment by refi93** [comment]
@matejcik done, but I found a bug in the previously merged PR regarding account indices beyond 100:

https://github.com/trezor/trezor-firmware/pull/1423/files#diff-eaac9c946526e07b3ac5c317971467dc330f41ee2cd7da91e701aa3098365bdbR729

There is an off by 1 error in the account index check and it compares the hardened account index against a non-hardened constant, so I'm pushing an additional fix commit to this PR (with a test case to capture it) with additional tests/minor constants refactor to prevent shooting ourselves in the foot in the future

So I'm re-requesting your review

File: core/tests/test_apps.cardano.sign_tx.py [hunk 1]
```diff
@@ -9,7 +9,7 @@
 @unittest.skipUnless(not utils.BITCOIN_ONLY, "altcoin")
 class TestCardanoSignTransaction(unittest.TestCase):
     def test_should_show_outputs(self):
-        outputs_to_show = [
+        outputs_to_hide = [
```
Comment (refi93) [inline core/tests/test_apps.cardano.sign_tx.py:12]: those tests apparently were written for a function called `_should_show_output` and were not updated correctly during the refactor to `_should_hide_output`, ending up with a contradictory naming

File: core/src/apps/cardano/helpers/paths.py [hunk 2]
```diff
@@ -18,10 +18,11 @@
 
 # the maximum allowed change address.  this should be large enough for normal
 # use and still allow to quickly brute-force the correct bip32 path
-MAX_CHANGE_ADDRESS_INDEX = const(1_000_000)
-MAX_ACCOUNT_INDEX = const(100)
+MAX_SAFE_CHANGE_ADDRESS_INDEX = const(1_000_000)
+MAX_SAFE_ACCOUNT_INDEX = const(100) | HARDENED
 ACCOUNT_PATH_INDEX = const(2)
-BIP_PATH_LENGTH = const(5)
+PURPOSE_PATH_INDEX = const(3)
```
Comment (matejcik) [inline core/src/apps/cardano/helpers/paths.py:24]: `purpose` in BIP-44 refers to the _first_ component.

The fourth is called `change`, which is not appropriate here. Maybe "address type" ? But that also means something else

File: core/src/apps/cardano/sign_tx.py [hunk 3]
```diff
@@ -724,11 +725,11 @@ def _should_hide_output(output: List[int], inputs: List[CardanoTxInputType]) ->
     for tx_input in inputs:
         inp = tx_input.address_n
         if (
-            len(output) != BIP_PATH_LENGTH
+            len(output) != ADDRESS_PATH_INDEX + 1
```
Comment (matejcik) [inline core/src/apps/cardano/sign_tx.py:728]: this is somewhat less readable. I'd rather keep the code as it was, except for the bugfix

File: core/src/apps/cardano/helpers/paths.py (hunk 2 above)
Comment (refi93) [reply core/src/apps/cardano/helpers/paths.py:24]: yes indeed, I mixed up the terminology here. The corresponding part of the path is actually called "chain" in Cardano terminology: https://github.com/cardano-foundation/CIPs/blob/master/CIP-1852/CIP-1852.md

**Comment by refi93** [comment]
@matejcik rebased and pushed suggested change (not touching existing constants not relevant to the bugfix)

**Comment by refi93** [comment]
@matejcik I rebased it again - is there any blocker preventing this PR from being merged we can help with?

**Comment by matejcik** [comment]
@refi93 no blocker, just a little backlog on our side. it's going in next week at the latest
